Structure_Engine/Compute/ReinforcementDensity.cs

```diff
@@ -0,0 +1,199 @@
+
+            foreach(Bar bar in bars)
+            {
+                IFragment id;
+                bar.Fragments.TryGetValue(fragmentType, out id);
+                ids.Add((((IAdapterId)id).Id.ToString()));
+            }
+
+            if(ids.Count == 0)
+            {
+                Reflection.Compute.RecordError("No ids were found within the Bars, therefore the ReinforcementDensity cannot be evaluated.");
+                return null;
+            }
+
+            Dictionary<string, Bar> barsDict = ids.Zip(bars, (k, v) => new { k, v }).ToDictionary(x => x.k, x => x.v);
+
+            IMaterialFragment material;
+            Bar resultBar;
+            ReinforcementDensity reinforcementDensity;
+
+            if(materialsDict.TryGetValue(barRequiredArea.MaterialName, out material) && barsDict.TryGetValue(barRequiredArea.ObjectId.ToString(), out resultBar))
+            {
+                //Calculate the volume of the Bar
+                if(resultBar.SectionProperty == null)
+                {
+                    Reflection.Compute.RecordError("The Bar defined in the BarRequiredArea does not have a SectionProperty. Therefore, the ReinforcementDensity cannot be evaluated.");
+                    return null;
+                }
+                double elementArea = resultBar.SectionProperty.Area;
+                double reinforcedArea = barRequiredArea.SumRequiredArea();
+                double density = material.Density;
+                double rebarDensity = reinforcedArea * density / elementArea;
+                reinforcementDensity = Create.ReinforcementDensity(rebarDensity, material);
+            }
+            else
+            {
+                Reflection.Compute.RecordError("The Bar and/or Material could not be found in the lists provided corresponding to the ids provided for in the BarRequiredArea.");
+                return null;
+            }
+
+            return reinforcementDensity;
+        }
+
+        /***************************************************/
+
+        [Description("Calculates the ReinforcementDensity of a MeshRequiredArea from a given Bar and Material.")]
+        [Input("meshRequiredArea", "The BarRequiredArea containing the reinforced areas and ids for the Bar and Material.")]
+        [Input("elements", "The elements to search for the relevant AreaElement associated with the MeshRequiredArea.")]
+        [Input("materials", "The Materials to search for the relevant Material associated with the MeshRequiredArea.")]
+        [Output("reinforcementDensity", "The ReinforcementDensity calculated using the inputs provided.")]
+        public static ReinforcementDensity ReinforcementDensity(MeshRequiredArea meshRequiredArea, List<IAreaElement> elements, List<IMaterialFragment> materials)
+        {
+            Reflection.Compute.RecordError("Requires an implementation of Regions before the required area can be extrapolated.");
+            return null;
+            //Add to IsNull when PR is merged
```

Comment:
I guess this is meant to stay here for the future, but just to double check - please make sure the code below is needed and should not be deleted.